lib/Reader/abisignature.cpp

```diff
@@ -436,7 +461,13 @@ Value *ABICallSignature::emitCall(GenIR &Reader, Value *Target, bool MayThrow,
       ResultNode = Call.getInstruction();
     }
   } else {
-    ResultNode = Builder.CreateLoad(ResultNode);
+    Type *ResultTy = ResultNode->getType();
+    if (!ResultTy->isPointerTy() ||
+        !ResultTy->getPointerElementType()->isStructTy()) {
+      ResultNode = Builder.CreateLoad(ResultNode);
+    } else {
+      Reader.setValueRepresentsStruct(ResultNode);
```

Comment:
I wonder if it might be better to move this up into the indirect result setup (see lines 351-354). If you call `GenIR::setValueRepresentsStruct` up there, this becomes

```
if (!Reader.doesValueRepresentStruct(ResultNode)) {
    ResultNode = Builder.CreateLoad(ResultNode);
}
```